File: ruby/09_from_push_to_pull/lib/values_of_month.rb
class ValuesOfMonth
  attr_reader :date_of_month, :transactions_of_month, :preceding_balance
  private :date_of_month, :transactions_of_month, :preceding_balance

  def initialize(date_of_month = Date.today, transactions_of_month = [], preceding_balance = 0)
    @date_of_month = date_of_month
    @transactions_of_month = transactions_of_month
    @preceding_balance = preceding_balance
  end

  def balance
    transactions_of_month.map(&:amount).inject(preceding_balance, :+)
  end

  def average_balance
    balance = preceding_balance
    ultimo = date_of_month.day

    average_balance = 0.0
    day_of_latest_balance = 1
    transactions_of_month.each do |transaction|
      day = transaction.date.day
      average_balance += calculate_proportional_balance(day_of_latest_balance, balance, day, ultimo)
      balance += transaction.amount
      day_of_latest_balance = day
    end
    average_balance += calculate_proportional_balance(day_of_latest_balance, balance, ultimo + 1, ultimo)
    Integer(average_balance)
  end

  private
  def calculate_proportional_balance(day_of_latest_balance, balance, day, days_in_month)
    counting_days = day - day_of_latest_balance
    return 0 if counting_days == 0
    rate = counting_days.to_f / days_in_month.to_f
    balance * rate
  end
end

File: csharp/09_FromPushToPull/ValuesOfMonth.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Common;

namespace FromPushToPull
{
    class ValuesOfMonth
    {
        private readonly DateTime dateOfMonth;
        private readonly IList<Transaction> transactionsOfMonth;
        private readonly int precedingBalance;

        public int Balance
        {
            get
            {
                int balance = precedingBalance;
                foreach (Transaction transaction in transactionsOfMonth)
                {
                    balance += transaction.Amount;
                }
                return balance;
            }
        }

        public int AverageBalance
        {
            get
            {
                int balance = precedingBalance;
                int ultimo = dateOfMonth.Day;

                double averageBalance = 0;
                int dayOfLatestBalance = 1;
                foreach (Transaction transaction in transactionsOfMonth)
                {
                    int day = transaction.Date.Day;
                    averageBalance += CalculateProportionalBalance(dayOfLatestBalance, balance, day, ultimo);
                    balance += transaction.Amount;
                    dayOfLatestBalance = day;
                }

                averageBalance += CalculateProportionalBalance(dayOfLatestBalance, balance, ultimo + 1, ultimo);

                return (int)averageBalance;
            }
        }

        public ValuesOfMonth() : this(DateTime.Now, new List<Transaction>(), 0) { }

        public ValuesOfMonth(DateTime dateOfMonth, IList<Transaction> transactionsOfMonth, int precedingBalance)
        {
            this.dateOfMonth = dateOfMonth;
            this.transactionsOfMonth = transactionsOfMonth;
            this.precedingBalance = precedingBalance;
        }

        private double CalculateProportionalBalance(int dayOfLatestBalance, int balance, int day, int daysInMonth)
        {
            int countingDays = day - dayOfLatestBalance;
            if (countingDays == 0)
            {
                return 0;
            }
            double rate = (double)countingDays / daysInMonth;
            return (balance * rate);
        }
    }
}
